### DISCUSSION TITLE
producer confirm_delivery

### DISCUSSION THREAD

**Discussion by edsionxuanxuan** [pika/pika#1445] [Q&A]
hi: 

version: 1.3.2
os: windows10
python: 3.10

I want to  ask , in the new version of pika, the producer confirm mode, producter-->exchange, exchange-->queue, only need to use try... except.. to catch busic_publish? If producter-->exchange fails, catch NackError, then exchange-->queue, catch UnroutableError?
old version:

```
def on_delivery_confirmation(method_frame):
    confirmation_type = method_frame.method.NAME
    if confirmation_type == 'Basic.Ack':
        print("Message published and confirmed!")
    else:
        print("Message not confirmed!")
channel.confirm_delivery()
channel.basic_publish(exchange='', routing_key='queue_name', body=message,
                      properties=pika.BasicProperties(content_type='text/plain'))
channel.add_on_return_callback(on_delivery_confirmation)
```

new version:
```
    channel.confirm_delivery()
    
    try:
        channel.basic_publish(exchange='test_exchange',
                routing_key='standard_key',
                body=json.dumps(message),
                properties=pika.BasicProperties(content_type='application/json', delivery_mode=DeliveryMode.Persistent))
        print('Message publish was confirmed')
    except pika.exceptions.UnroutableError:
        print('exchange routing Message to queue could not be confirmed')
    except pika.exceptions.NackError:
        print("product send Message to exchange could not be confirmed ")
```

thanks

**Comment by lukebakken** [comment]
Hello, could you please provide a git repository with **complete** code I can run?

**Reply by kushalgithub18** [reply]
hey,
so there is no concept of basic.ack and basic.nack in the newer version in the publisher confirms?

**Reply by lukebakken** [reply]
> so there is no concept of basic.ack and basic.nack in the newer version in the publisher confirms?

What makes you think that? Of course this library supports `basic.ack` and `basic.nack`.
